gym/spaces/dict.py

```diff
@@ -83,8 +83,17 @@ def __init__(
             spaces = spaces_kwargs
         if isinstance(spaces, dict) and not isinstance(spaces, OrderedDict):
             try:
-                spaces = OrderedDict(sorted(spaces.items()))
-            except TypeError:  # raise when sort by different types of keys
+                spaces = OrderedDict(
+                    # Add `key.__class__.__qualname__` to support sorting between
+                    # different types (e.g. `int` vs. `str`)
+                    sorted(
```

Comment:
I still don't like this, it actively restricts usage for no good reason, and is done via an extremely hacky method. Imo this should be kept as it was earlier, and the only change in this PR would be the `__eq__` method.

Reply:
The semantic to change the `key` argument in the `sorted` function is to make:

Equal `dict`s imply equal `Dict` spaces output.

```python
dict1 == dict2 <==> spaces.Dict(dict1) == spaces.Dict(dict2)
```

If we don't allow sorting with different types but just update the `__eq__` method, we will get:

```python
# Python dict
{1: subspace1, 'a': subspace2} == {'a': subspace2, 1: subspace1}
```

but

```python
Dict(1: subspace1, 'a': subspace2) != Dict('a': subspace2, 1: subspace1)
```

Note that if the users always pass the keys in the same type, the new implementation here behaves just like the old one.

------

If the users want to get full control of the key order in the `spaces.Dict`, they should pass the argument as `OrderedDict`:

```python
spaces.Dict(spaces=OrderedDict([
    (key1, subspace1),
    ...
    (keyN, subspaceN),
]))
```

We only sort keys if the argument `spaces` is passed by vanilla Python `dict` or unordered `Mapping`. Which are:

```python
spaces.Dict(spaces={key1: subspace1, ..., keyN: subspaceN})  # arbitrary key type
spaces.Dict(key1=subspace1, ..., keyN=subspaceN)  # keys are `str` and sortable
```

The latter one is always sortable because the keys are `str`s.

Reply:
Im uncertain about this PR. 
The downside is that this isn't backwards compatible which is only in a couple of very niche cases.
The upside is that this fixes a bug that could occur in those niche cases. 

I think we have two options.
1. We explicitly don't support multi different types of keys for Dict spaces, therefore won't need the PR
2. Support multi different types and accept this PR.

@XuehaiPan In what cases do the multi different types of keys occur? The only way that I can think of this occurring is by using custom objects as the keys.